### ISSUE TITLE
Reuse of layers with mixed precision produce multiple float16 copies of weights

### ISSUE THREAD

**Issue by tringholm** [keras-team/keras#15474]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Arch Linux 5.12.12
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.6.0
- Python version: 3.8.7
- GPU model and memory: NVIDIA GeForce GTX 1050, 3 GB

**Describe the problem**.

Keras produces multiple float16 copies of casted float32 variables when reusing layers with mixed precision, causing OOM issues.

**Describe the current behavior**.

Keras AutoCastVariables are casted from float32 to float16 on use when the mixed precision policy is active. If a layer with mixed precision policy is called multiple times (as is the case for RNN cells, for instance), every usage of the layer's weights produces a separate float16 copy of the master float32 weights. 

With repeated activation, this copy-on-use behaviour can inflate memory usage, causing OOM issues.

**Describe the expected behavior**.

Expected behaviour is that layers should produce a single float16 copy of weights for use in computation, instead of producing a copy for every operation.

**Standalone code to reproduce the issue**.
The following example code is unrelated to the code where the issue was first discovered, and is not expected to produce a sensible model. By varying num_iterations to e.g. 50, one can see that memory usage balloons far beyond what's expected for 40 additional activations.

```
import tensorflow as tf
import numpy as np
from tensorflow.keras import Model
from tensorflow.keras.layers import Input, Dense
from tensorflow.keras.optimizers import Adam

policy = tf.keras.mixed_precision.Policy('mixed_float16')
tf.keras.mixed_precision.experimental.set_policy(policy)

gpus = tf.config.experimental.list_physical_devices('GPU')
for gpu in gpus:
    tf.config.experimental.set_memory_growth(gpu, True)

layer_dim = 5000
num_examples = 100
num_iterations = 10

dense_layer = Dense(layer_dim)

X = np.random.normal(size=num_examples * layer_dim).reshape(num_examples, layer_dim, 1)
Y = np.zeros(shape=(num_examples, layer_dim, 1))

inputs = Input(shape=layer_dim)
x = dense_layer(inputs)
for i in range(num_iterations):
    x = dense_layer(x)
outputs = dense_layer(x)

train_model = Model(inputs=inputs, outputs=outputs)
train_model.compile(optimizer=Adam(), loss='mean_squared_error')
train_model.summary()

train_model.fit(x=X, y=Y, epochs=1, shuffle=False)
```

**Source code / logs**.

I am not at liberty to provide actual logs from the model that first produced the issue, but what put me on to this issue was seeing logs like

"tensorflow/core/common_runtime/bfc_allocator.cc:456] Allocator (GPU_0_bfc) ran out of memory trying to allocate 2.08MiB (rounded to 2179072)requested by op ...",
"Current allocation summary follows.",
"tensorflow/core/common_runtime/bfc_allocator.cc:991] BFCAllocator dump for GPU_0_bfc",
"InUse at 7f3166000000 of size 2097152 by op DECODER/while/RNNCELL/MatMul/Cast ",

where, upon further inspection, Cast operations were being done at every activation of the RNN cell.

**Comment by reedwm** [comment]
I agree this is inconvenient, but unfortunately there's currently no solution currently to cache the casted value of AutoCastVariable. This is a tricky problem to solve, since a layer can modify the value of the weights in `call`, which would invalid the cached value if we did implement caching.

Another workaround for subclassed layers is to create the variable in `OuterLayer`, call `variable.read_value()` in `OuterLayer.call` (which causes the variable to be automatically casted), then pass the value to `InnerLayer.call`. This way the variable is read only once, so it is only casted once.

For some of the built-in RNN classes, the [variable is read outside the loop](https://github.com/keras-team/keras/blob/f5171d521acbf2ebbb6414352d5792163c41479f/keras/layers/recurrent_v2.py#L1194), so that the variable is casted only once per step instead of once per iteration of the RNN. For custom RNNs, currently you would need to implement that logic yourself to avoid creating many float16 copies of the variable.

**Comment by SuryanarayanaY** [comment]
Hello, Thank you for reporting an issue.

We're currently in the process of migrating the new Keras 3 code base from keras-team/keras-core to keras-team/keras.
Consequently, This issue may not be relevant to the Keras 3 code base . After the migration is successfully completed, feel free to reopen this Issue at keras-team/keras if you believe it remains relevant to the Keras 3 code base. If instead this Issue is a bug or security issue in legacy tf.keras, you can instead report a new issue at [keras-team/tf-keras](https://github.com/keras-team/tf-keras/issues), which hosts the TensorFlow-only, legacy version of Keras.

To know more about Keras 3, please read https://keras.io/keras_core/announcement/
